File: Diagrammatize.Design.Pattern/code/cxx/AbstractFactory/A2/listfactory/ListFactory.cpp
#include "ListFactory.hpp"
#include "ListLink.hpp"
#include "ListTray.hpp"
#include "ListPage.hpp"

Link* ListFactory::createLink(const std::string& caption, const std::string& url) {
    return new ListLink(caption, url);
}

Tray* ListFactory::createTray(const std::string& caption) {
    return new ListTray(caption);
}

Page* ListFactory::createPage(const std::string& title, const std::string& author) {
    return new ListPage(title, author);
}

static Factory* newListFactory() {
    return new ListFactory();
}

//static FactoryRegister factory("ListFactory", &newListFactory);
REGISTER_FACTORY(ListFactory, &newListFactory);

File: Diagrammatize.Design.Patterns/code/python/AbstractFactory/A2/listfactory/ListFactory.py
from factory import Factory
from .ListLink import ListLink
from .ListTray import ListTray
from .ListPage import ListPage


class ListFactory(Factory):
    def createLink(self, caption, url):
        return ListLink(caption, url)

    def createTray(self, caption):
        return ListTray(caption)

    def createPage(self, title, author):
        return ListPage(title, author)
